Use fixed-size deadCells vector in MarkedBlock::sweep

ddegazio opened the pull request:
#### 424d1996eedf829b94d9593ed1a4288994bf43f1
<pre>
Use fixed-size deadCells vector in MarkedBlock::sweep
<a href="https://bugs.webkit.org/show_bug.cgi?id=254960">https://bugs.webkit.org/show_bug.cgi?id=254960</a>
rdar://107585451

Reviewed by Justin Michaud and Yusuke Suzuki.

Allocates the deadCells vector in MarkedBlock::specializedSweep
entirely on the stack, and stores atom numbers using uint16_t instead
of uint64_t. Should reduce allocations and potentially improve locality.

* Source/JavaScriptCore/heap/MarkedBlockInlines.h:
(JSC::MarkedBlock::Handle::specializedSweep):

Canonical link: <a href="https://commits.webkit.org/262612@main">https://commits.webkit.org/262612@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/db14d18b0ea8c5a6c0509b32590fc4d0417f66d2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2020 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2049 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2111 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2944 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2085 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2125 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2089 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1848 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2037 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1814 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1810 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2791 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1801 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1790 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1702 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/1689 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1855 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1838 "5 failures") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2964 "267 api tests failed or timed out") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/1926 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1854 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1658 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/2078 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1779 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1801 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/480 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/507 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1961 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/2123 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/595 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (approved):
r=me

Constellation reviewed (approved):
r=me

tadeuzagallo commented on Source/JavaScriptCore/heap/MarkedBlockInlines.h:
```diff
@@ -309,7 +309,9 @@ void MarkedBlock::Handle::specializedSweep(FreeList* freeList, MarkedBlock::Hand
     bool isEmpty = true;
-    Vector<size_t> deadCells;
+    static_assert(UINT16_MAX <= atomsPerBlock);
+    static_assert(atomsPerBlock <= 2048); // Arbitrary size limit so the vector doesn't get too big.
```
These two things can't be true at the same time, right?

ddegazio replied:
Oh whoops! That first one is reversed, good catch. I need to adjust the other assertion too to support platforms with 64k pages, so I'll push an amendment shortly.

ddegazio committed:
Use fixed-size deadCells vector in MarkedBlock::sweep
https://bugs.webkit.org/show_bug.cgi?id=254960
rdar://107585451

Reviewed by Justin Michaud and Yusuke Suzuki.

Allocates the deadCells vector in MarkedBlock::specializedSweep
entirely on the stack, and stores atom numbers using uint16_t instead
of uint64_t. Should reduce allocations and potentially improve locality.

* Source/JavaScriptCore/heap/MarkedBlockInlines.h:
(JSC::MarkedBlock::Handle::specializedSweep):

Canonical link: https://commits.webkit.org/262612@main
